File: android/app/src/main/kotlin/dev/agentdeck/util/SessionGrouping.kt
package dev.agentdeck.util

/**
 * Project-prefix session grouping — Kotlin mirror of
 * `shared/src/session-utils.ts` (`normalizeProjectForGrouping` /
 * `projectGroupKey` / `groupSessionsByProject`), itself a port of the IPS10
 * terrarium huddle builder (esp32/src/ui/terrarium/office.cpp
 * `normProject` / `sameProjectGroup`). Keep the three in lockstep.
 *
 * Sessions whose project names share a long delimiter-aligned prefix
 * (worktree/task folders like `xteink-x3-x4-japanese-broken-claude-glm` /
 * `…-broken-codex`) cluster into one work group; short siblings like
 * `agentdeck-ios` / `agentdeck-android` stay separate.
 */

private fun isProjectDelim(c: Char): Boolean =
    c == '-' || c == '_' || c == ' ' || c == '.'

private fun trimProjectTail(s: String): String {
    var n = s.length
    while (n > 0 && (isProjectDelim(s[n - 1]) || s[n - 1] == '#')) n--
    return s.substring(0, n)
}

private fun projectDelimCount(s: String, len: Int): Int {
    var count = 0
    for (i in 0 until minOf(len, s.length)) if (isProjectDelim(s[i])) count++
    return count
}

/** Basename + strip trailing " #N" duplicate suffix. */
fun normalizeProjectForGrouping(project: String?): String {
    val raw = project?.trim().orEmpty()
    val base = raw.split('/').lastOrNull { it.isNotEmpty() } ?: raw
    val m = Regex("""^(.*?)\s*#\d+$""").find(base)
    return trimProjectTail(m?.groupValues?.get(1) ?: base)
}

/** Shared group key (common stem) when [a] and [b] belong to the same work
 *  group, else null. Conservative: stem >= 14 chars with >= 2 delimiters on
 *  both sides, or an exact case-insensitive match. */
fun projectGroupKey(a: String, b: String): String? {
    var i = 0
    var lastDelim = -1
    val n = minOf(a.length, b.length)
    while (i < n && a[i].lowercaseChar() == b[i].lowercaseChar()) {
        if (isProjectDelim(a[i])) lastDelim = i
        i++
    }
    if (i == a.length && i == b.length) return a

    val stemLen = when {
        i == a.length && i < b.length && isProjectDelim(b[i]) -> i
        i == b.length && i < a.length && isProjectDelim(a[i]) -> i
        lastDelim > 0 -> lastDelim
        else -> -1
    }
    if (stemLen < 14 || projectDelimCount(a, stemLen) < 2 || projectDelimCount(b, stemLen) < 2) {
        return null
    }
    val key = trimProjectTail(a.substring(0, stemLen))
    return key.ifEmpty { null }
}

data class ProjectGroup<T>(
    /** Group label — shared stem, or the normalized project name for singletons. */
    val key: String,
    /** True when >= 2 members fused (render a group header); singletons render flat. */
    val grouped: Boolean,
    val members: List<T>,
)

/** Cluster an ordered list into project groups, preserving input order. */
fun <T> groupSessionsByProject(items: List<T>, projectOf: (T) -> String?): List<ProjectGroup<T>> {
    val groups = mutableListOf<ProjectGroup<T>>()
    val done = BooleanArray(items.size)
    for (a in items.indices) {
        if (done[a]) continue
        val na = normalizeProjectForGrouping(projectOf(items[a]))
        var key = na
        val members = mutableListOf(items[a])
        done[a] = true
        for (b in a + 1 until items.size) {
            if (done[b]) continue
            val nb = normalizeProjectForGrouping(projectOf(items[b]))
            val nextKey = projectGroupKey(key, nb) ?: projectGroupKey(na, nb)
            if (nextKey != null) {
                key = nextKey
                members.add(items[b])
                done[b] = true
            }
        }
        groups.add(ProjectGroup(key = key, grouped = members.size > 1, members = members))
    }
    return groups
}

File: apple/AgentDeck/Model/SessionGrouping.swift
// SessionGrouping.swift — Project-prefix session grouping.
//
// Swift mirror of `shared/src/session-utils.ts` (`normalizeProjectForGrouping`
// / `projectGroupKey` / `groupSessionsByProject`), itself a port of the IPS10
// terrarium huddle builder (esp32/src/ui/terrarium/office.cpp `normProject` /
// `sameProjectGroup`). Kotlin mirror: android `dev.agentdeck.util.SessionGrouping`.
// Keep all three in lockstep.
//
// Sessions whose project names share a long delimiter-aligned prefix
// (worktree/task folders like `xteink-x3-x4-japanese-broken-claude-glm` /
// `…-broken-codex`) cluster into one work group; short siblings like
// `agentdeck-ios` / `agentdeck-android` stay separate.

import Foundation

enum SessionGrouping {
    private static func isDelim(_ c: Character) -> Bool {
        c == "-" || c == "_" || c == " " || c == "."
    }

    private static func trimTail(_ s: String) -> String {
        let chars = Array(s)
        var n = chars.count
        while n > 0 && (isDelim(chars[n - 1]) || chars[n - 1] == "#") { n -= 1 }
        return String(chars[0..<n])
    }

    private static func delimCount(_ s: [Character], upTo len: Int) -> Int {
        var count = 0
        for i in 0..<min(len, s.count) where isDelim(s[i]) { count += 1 }
        return count
    }

    /// Basename + strip trailing " #N" duplicate suffix.
    static func normalizeProject(_ project: String?) -> String {
        let raw = (project ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        let base = raw.split(separator: "/").last.map(String.init) ?? raw
        if let range = base.range(of: #"\s*#\d+$"#, options: .regularExpression) {
            return trimTail(String(base[..<range.lowerBound]))
        }
        return trimTail(base)
    }

    /// Shared group key (common stem) when `a` and `b` belong to the same work
    /// group, else nil. Conservative: stem >= 14 chars with >= 2 delimiters on
    /// both sides, or an exact case-insensitive match.
    static func groupKey(_ a: String, _ b: String) -> String? {
        let ca = Array(a), cb = Array(b)
        var i = 0
        var lastDelim = -1
        let n = min(ca.count, cb.count)
        while i < n, String(ca[i]).lowercased() == String(cb[i]).lowercased() {
            if isDelim(ca[i]) { lastDelim = i }
            i += 1
        }
        if i == ca.count && i == cb.count { return a }

        var stemLen = -1
        if i == ca.count && i < cb.count && isDelim(cb[i]) { stemLen = i }
        else if i == cb.count && i < ca.count && isDelim(ca[i]) { stemLen = i }
        else if lastDelim > 0 { stemLen = lastDelim }

        if stemLen < 14 || delimCount(ca, upTo: stemLen) < 2 || delimCount(cb, upTo: stemLen) < 2 {
            return nil
        }
        let key = trimTail(String(ca[0..<stemLen]))
        return key.isEmpty ? nil : key
    }

    struct Group<T> {
        /// Group label — shared stem, or the normalized project name for singletons.
        let key: String
        /// True when >= 2 members fused (render a group header); singletons render flat.
        let grouped: Bool
        let members: [T]
    }

    /// Cluster an ordered list into project groups, preserving input order.
    static func group<T>(_ items: [T], projectOf: (T) -> String?) -> [Group<T>] {
        var groups: [Group<T>] = []
        var done = [Bool](repeating: false, count: items.count)
        for a in items.indices where !done[a] {
            let na = normalizeProject(projectOf(items[a]))
            var key = na
            var members = [items[a]]
            done[a] = true
            for b in (a + 1)..<items.count where !done[b] {
                let nb = normalizeProject(projectOf(items[b]))
                if let nextKey = groupKey(key, nb) ?? groupKey(na, nb) {
                    key = nextKey
                    members.append(items[b])
                    done[b] = true
                }
            }
            groups.append(Group(key: key, grouped: members.count > 1, members: members))
        }
        return groups
    }
}
